bench/lib/benchmark.js

```diff
@@ -0,0 +1,93 @@
+// @flow
+
+'use strict';
+
+class Benchmark {
+    constructor() {
+        this._async = this._async.bind(this);
+    }
+
+    /**
+     * The `setup` method is intended to be overridden by subclasses. It will be called once, prior to
+     * running any benchmark iterations, and may set state on `this` which the benchmark later accesses.
+     * If the setup involves an asynchronous step, `setup` may return a promise.
+     */
+    setup(): Promise<void> | void {}
+
+    /**
+     * The `bench` method is intended to be overridden by subclasses. It should contain the code to be
+     * benchmarked. It may access state on `this` set by the `setup` function (but should not modify this
+     * state). It will be called multiple times, the total number to be determined by the harness. If
+     * the benchmark involves an asynchronous step, `bench` may return a promise.
+     */
+    bench(): Promise<void> | void {}
+
+    /**
+     * The `teardown` method is intended to be overridden by subclasses. It will be called once, after
+     * running all benchmark iterations, and may perform any necessary cleanup. If cleaning up involves
+     * an asynchronous step, `teardown` may return a promise.
+     */
+    teardown(): Promise<void> | void {}
+
+    _async: () => Promise<Array<number>>;
+    _elapsed: number;
+    _samples: Array<number>;
+    _start: number;
+
+    /**
+     * Run the benchmark by executing `setup` once, sampling the execution time of `bench` some number of
+     * times, and then executing `teardown`. Yields an array of execution times.
+     */
+    run(): Promise<Array<number>> {
+        return Promise.resolve(this.setup()).then(() => this._begin());
+    }
+
+    _done() {
+        // 210 samples => 20 observations for regression
+        return this._elapsed >= 300 && this._samples.length > 210;
+    }
+
+    _begin(): Promise<Array<number>> {
+        this._samples = [];
+        this._elapsed = 0;
+        this._start = performance.now();
+
+        const bench = this.bench();
+        if (bench instanceof Promise) {
+            return bench.then(this._async);
+        } else {
+            return (this._sync(): any);
+        }
+    }
+
+    _sync() {
+        // Avoid Promise overhead for sync benchmarks.
+        while (true) {
+            const sample = performance.now() - this._start;
+            this._samples.push(sample);
+            this._elapsed += sample;
+            if (this._done()) {
+                return this._end();
+            }
+            this._start = performance.now();
+            this.bench();
+        }
+    }
```

Comment:
Noticed that this locks up the UI, with the currently-running benchmark below the fold so that it's not immediately obvious why the page is "frozen".   Not a huge deal, but it might be a slightly better experience if each benchmark's results were collapsed by default, except for the currently-running one.

Reply:
Let's keep them expanded for now, so that if you use an extension to snapshot the page you get everything by default. I did add a "Running" / "Finished" indicator to the top of the page so you can see without scrolling if it's done or not. I anticipate the most common usage will be either:

* Focused on one benchmark (using URL hash)
* Running all benchmarks (while getting coffee/lunch)